From 5fc41fae9300b5ea9ca685f33c3cdf1d719c791f Mon Sep 17 00:00:00 2001
From: lrinQVQ <cqlrin@gmail.com>
Date: Wed, 17 Oct 2018 09:15:36 +0800
Subject: [PATCH 1/4] Core: add function to decode hexadecimal strings.

Adds functionality to convert a hexadecimal string into binary data.
This will be used to decode PSKs stored in hexadecimal representation.

Author:     Nate Karstens <nate.karstens@garmin.com>
Date:      Wed Aug 23 21:00:18 2017 -0500
---
 src/core/ngx_string.c | 50 +++++++++++++++++++++++++++++++++++++++++++
 src/core/ngx_string.h |  1 +
 2 files changed, 51 insertions(+)

diff --git a/src/core/ngx_string.c b/src/core/ngx_string.c
index 2ee07bfc..20d95451 100644
--- a/src/core/ngx_string.c
+++ b/src/core/ngx_string.c
@@ -1134,6 +1134,56 @@ ngx_hex_dump(u_char *dst, u_char *src, size_t len)
 }
 
 
+ngx_int_t
+ngx_hex_decode(u_char *dst, u_char *src, size_t len)
+{
+    u_char  ch, decoded;
+
+    if (len & 1) {
+        return NGX_ERROR;
+    }
+
+    while (len) {
+        ch = *src++;
+        len -= 2;
+
+        if (ch >= '0' && ch <= '9') {
+            decoded = ch - '0';
+            goto second;
+        }
+
+        ch |= 0x20;
+
+        if (ch >= 'a' && ch <= 'f') {
+            decoded = ch - 'a' + 10;
+            goto second;
+        }
+
+        return NGX_ERROR;
+
+    second:
+
+        ch = *src++;
+
+        if (ch >= '0' && ch <= '9') {
+            *dst++ = (u_char) ((decoded << 4) + ch - '0');
+            continue;
+        }
+
+        ch |= 0x20;
+
+        if (ch >= 'a' && ch <= 'f') {
+            *dst++ = (u_char) ((decoded << 4) + ch - 'a' + 10);
+            continue;
+        }
+
+        return NGX_ERROR;
+    }
+
+    return NGX_OK;
+}
+
+
 void
 ngx_encode_base64(ngx_str_t *dst, ngx_str_t *src)
 {
diff --git a/src/core/ngx_string.h b/src/core/ngx_string.h
index 882ae7ce..e3a93d6b 100644
--- a/src/core/ngx_string.h
+++ b/src/core/ngx_string.h
@@ -179,6 +179,7 @@ time_t ngx_atotm(u_char *line, size_t n);
 ngx_int_t ngx_hextoi(u_char *line, size_t n);
 
 u_char *ngx_hex_dump(u_char *dst, u_char *src, size_t len);
+ngx_int_t ngx_hex_decode(u_char *dst, u_char *src, size_t len);
 
 
 #define ngx_base64_encoded_length(len)  (((len + 2) / 3) * 4)
-- 
2.17.1

